File: recipe-ts-runtime/src/DirectDispatchOven.ts
import { Dispatcher, Oven } from "./Oven";
import { BackendOven } from "./BackendOven";

export class DirectDispatchOven extends Oven {
    constructor(readonly backendOven: BackendOven) {
        super();
        super._setDefaultDispatcher(payload => Promise.resolve(backendOven.bake(payload)));
    }

    public addDispatcher(domain: string, dispatcher: Dispatcher) {
        throw new Error("cannot add dispatchers to a DirectDispatchOven");
    }

    public setDefaultDispatcher(dispatcher?: Dispatcher) {
        throw new Error("cannot overwrite default dispatcher in DirectDispatchOven");
    }
}

File: recipe-java-runtime/src/main/java/ca/derekcormier/recipe/DirectDispatchOven.java
package ca.derekcormier.recipe;

public class DirectDispatchOven extends Oven {
  private BackendOven backendOven;

  public DirectDispatchOven(BackendOven backendOven) {
    _setDefaultDispatcher(payload -> backendOven.bake(payload));
  }

  @Override
  public void setDefaultDispatcher(Dispatcher dispatcher) {
    throw new UnsupportedOperationException(
        "cannot overwrite default dispatcher in DirectDispatchOven");
  }

  @Override
  public void addDispatcher(String domain, Dispatcher dispatcher) {
    throw new UnsupportedOperationException("cannot add dispatchers to a DirectDispatchOven");
  }
}
